packages/no-trapping-shim/src/no-trapping-pony.js

```diff
@@ -0,0 +1,206 @@
+  isNoTrapping(target) {
+    if (isPrimitive(target)) {
+      throw TypeError('Reflect.isNoTrapping called on non-object');
+    }
+    return isNoTrappingInternal(target, false);
+  },
+  suppressTrapping(target) {
+    if (isPrimitive(target)) {
+      throw TypeError('Reflect.suppressTrapping called on non-object');
+    }
+    return suppressTrappingInternal(target, false);
+  },
+});
+
+export const extraObjectMethods = freeze({
+  isNoTrapping(target) {
+    if (isPrimitive(target)) {
+      return true;
+    }
+    return isNoTrappingInternal(target, true);
+  },
+  suppressTrapping(target) {
+    if (isPrimitive(target)) {
+      return target;
+    }
+    if (suppressTrappingInternal(target, true)) {
+      return target;
+    }
+    throw TypeError('preventExtensions trap returned falsy');
+  },
+});
+
+const addExtras = (base, ...extrasArgs) => {
+  for (const extras of extrasArgs) {
+    for (const key of ownKeys(extras)) {
+      if (base[key] !== extras[key]) {
+        defineProperty(base, key, {
+          value: extras[key],
+          writable: true,
+          enumerable: false,
+          configurable: true,
+        });
+      }
+    }
+  }
+};
+
+const ReflectPlus = {};
+addExtras(ReflectPlus, OriginalReflect, extraReflectMethods);
+export { ReflectPlus };
+
+const ObjectPlus = function Object(...args) {
+  if (new.target) {
+    return construct(OriginalObject, args, new.target);
+  } else {
+    return apply(OriginalObject, this, args);
+  }
+};
+ObjectPlus.prototype = OriginalObject.prototype;
+addExtras(ObjectPlus, OriginalObject, extraObjectMethods);
+export { ObjectPlus };
+
+const makeMetaHandler = handler =>
+  freeze({
+    get(_, trapName, _receiver) {
+      return freeze((target, ...rest) => {
+        if (
+          isNoTrappingInternal(target, true) ||
+          handler[trapName] === undefined
+        ) {
+          return ReflectPlus[trapName](target, ...rest);
+        } else {
+          return handler[trapName](target, ...rest);
+        }
+      });
+    },
+  });
+
+const makeSafeHandler = handler =>
+  new OriginalProxy({}, makeMetaHandler(handler));
+
+/**
+ * In the shim, `ProxyPlus` should replace the global `Proxy`.
+ *
+ * @param {any} target
+ * @param {object} handler
+ */
+const ProxyPlus = function Proxy(target, handler) {
+  if (new.target !== ProxyPlus) {
+    if (new.target === undefined) {
+      throw TypeError('Proxy constructor requires "new"');
+    }
+    throw TypeError('Safe Proxy shim does not support subclassing');
+  }
+  const safeHandler = makeSafeHandler(handler);
+  const proxy = new OriginalProxy(target, safeHandler);
+  proxyHandlerMap.set(proxy, [target, handler]);
+  return proxy;
+};
+ProxyPlus.revocable = (target, handler) => {
+  const safeHandler = makeSafeHandler(handler);
+  const { proxy, revoke } = OriginalProxy.revocable(target, safeHandler);
+  proxyHandlerMap.set(proxy, [target, handler]);
+  return { proxy, revoke };
```

Comment:
I thought in our previous discussions we arrived at the conclusion that a non trapping proxy effectively cannot be revoked. I guess this should be an open issue for the proposal.

Reply:
Actually, our previous conclusion is necessary in order for, for example, a copyRecord to have the data-only stability properties we need. So no need to make this an open issue.

Done.